packages/next/src/build/webpack/loaders/lightningcss-loader/src/loader.ts

```diff
@@ -0,0 +1,309 @@
+import valueParser from 'next/dist/compiled/postcss-value-parser'
+import type { LoaderContext } from 'webpack'
+import type { ILightningCssLoaderConfig, VisitorOptions } from './interface'
+import { ECacheKey } from './interface'
+import { transform as transformCss, type Url, type Visitor } from 'lightningcss'
+import { Buffer } from 'buffer'
+import { getTargets } from './utils'
+import path from 'path'
+import {
+  getImportCode,
+  type ApiParam,
+  type ApiReplacement,
+  type CssExport,
+  type CssImport,
+  getModuleCode,
+  getExportCode,
+} from './codegen'
+import {
+  getFilter,
+  getPreRequester,
+  isDataUrl,
+  isUrlRequestable,
+  resolveRequests,
+} from '../../css-loader/src/utils'
+import { stringifyRequest } from '../../../stringify-request'
+
+function createVisitor(
+  options: ILightningCssLoaderConfig,
+  visitorOptions: VisitorOptions,
+  apis: ApiParam[],
+  imports: CssImport[],
+  replacements: ApiReplacement[]
+): Visitor<{}> {
+  const importUrlToNameMap = new Map<string, string>()
+
+  let hasUrlImportHelper = false
+  const urlToNameMap = new Map()
+  const urlToReplacementMap = new Map()
+  let urlIndex = -1
+
+  function handleUrl(u: Url): Url {
+    let url = u.url
+    const needKeep = visitorOptions.urlFilter(url)
+
+    if (!needKeep) {
+      return u
+    }
+
+    if (isDataUrl(url)) {
+      return u
+    }
+
+    const { urlResolver, context } = visitorOptions
+
+    const resolvedUrl = resolveRequests(urlResolver, context, [
```

Comment:
@devongovett Is there a way to `await` a promise from a visitor method?

Reply:
unfortunately not, and making the whole thing async would be a lot of work. Might add async visitor capabilities at some point but probably won't have time soon. What Vite and Parcel do currently is replace the dependencies with placeholder strings (hashes), then resolve all of the dependencies after the transform pass and replace the hashes with the final result. Maybe that could work here too?